engine/orchestration/src/com/cloud/vm/VirtualMachineManagerImpl.java

```diff
@@ -2744,7 +2748,11 @@ private void orchestrateMigrateAway(final String vmUuid, final long srcHostId, f
                 if (s_logger.isDebugEnabled()) {
                     s_logger.debug("Unable to find destination for migrating the vm " + profile);
                 }
-                throw new InsufficientServerCapacityException("Unable to find a server to migrate to.", host.getClusterId());
+                if (firstAttempt) {
+                    throw new InsufficientServerCapacityException("Unable to find a server to migrate to.", host.getClusterId());
+                } else {
+                    break;
+                }
             }
 
             excludes.addHost(dest.getHost().getId());
```

Comment:
you mentioned:

> Unfortunately we can't use the excludes list to know if this is the first attempt to migrate or not. The planDeployment method modifies the exclude list as it tries things internally.

Thoughts on why this line is even here? (line 2758)

are you saying this code modifies the excludes object?
`dest = _dpMgr.planDeployment(profile, plan, excludes, planner);`

Reply:
The `_dpMgr.planDeployment()` does modify the excludes. My cluster draining logic was added to there actually. Now that the loop is correctly looping, line 2758 prevents it from attempting to migrate to the same host over and over again. Otherwise it would likely just try the same host repeatedly.

Reply:
Thanks, code looks good and that all makes sense